codebase2/codebase-sqlite/U/Codebase/Sqlite/Sync22.hs

```diff
@@ -361,22 +361,27 @@ trySync tCache hCache oCache cCache = \case
       srcParents <- runSrc $ Q.loadCausalParents chId
       traverse syncCausal srcParents
 
+    -- Sync any watches of the given kinds to the dest if and only if watches of those kinds
+    -- exist in the src.
     syncWatch :: WK.WatchKind -> Sqlite.Reference.IdH -> m (TrySyncResult Entity)
     syncWatch wk r | debug && trace ("Sync22.syncWatch " ++ show wk ++ " " ++ show r) False = undefined
     syncWatch wk r = do
-      r' <- traverse syncHashLiteral r
-      doneKinds <- runDest (Q.loadWatchKindsByReference r')
-      if (notElem wk doneKinds) then do
-        runSrc (Q.loadWatch wk r) >>= traverse \blob -> do
-          TL.SyncWatchResult li body <-
-            either (throwError . DecodeError ErrWatchResult blob) pure $ runGetS S.decomposeWatchFormat blob
-          li' <- bitraverse syncTextLiteral syncHashLiteral li
-          when debug $ traceM $ "LocalIds for Source watch result " ++ show r ++ ": " ++ show li
-          when debug $ traceM $ "LocalIds for Dest watch result " ++ show r' ++ ": " ++ show li'
-          let blob' = runPutS $ S.recomposeWatchFormat (TL.SyncWatchResult li' body)
-          runDest (Q.saveWatch wk r' blob')
-        pure Sync.Done
-      else pure Sync.PreviouslyDone
+      runSrc (Q.loadWatch wk r) >>= \case
```

Comment:
The changes here aren't strictly necessary for the main goal of not syncing watch expressions, but I noticed that the old version first syncs the hash literal before even checking if the hash is a watch expression of the type we want to sync. This new version first checks that a watch expression of the desired type exists before we sync anything for it.